csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
I like the idea to check for null only for value types. Let's ignore `int?` `float?` etc for now.

I will amend the PR asap (probably this weekend) but I have a bigger picture in mine : have two repeatedfield implementations based on type (one for value types the other for ref types). A factory static method could take care of creating the correct implementation... Or maybe this is a little bit over engineered ...